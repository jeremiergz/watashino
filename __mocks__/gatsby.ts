import Gatsby from 'gatsby';

module.exports = {
  ...Gatsby,
  graphql: jest.fn(),
  navigate: jest.fn(),
  StaticQuery: jest.fn(),
  useStaticQuery: jest.fn(),
};
