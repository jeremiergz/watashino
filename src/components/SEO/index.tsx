import { graphql, useStaticQuery } from 'gatsby';
import React from 'react';
import { Helmet } from 'react-helmet-async';

const SEO: React.FC<SEOProps> = ({ description = '', keywords = [], lang = 'en', path = '', title = 'My Website' }) => {
  const {
    banner,
    site: { siteMetadata },
  } = useStaticQuery(graphql`
    query SEOData {
      banner: file(relativePath: { eq: "cover.png" }) {
        childImageSharp {
          fixed(height: 630, width: 1200, pngQuality: 100) {
            src
          }
        }
      }
      site {
        siteMetadata {
          description
          siteUrl
          title
          twitterUsername
        }
      }
    }
  `);
  const fullTitle = `${title} | ${siteMetadata.title}`;
  const metaDescription = description || siteMetadata.description;
  const siteUrl = siteMetadata.siteUrl;
  const imageUrl = `${siteUrl}${banner.childImageSharp.fixed.src}`;
  const url = `${siteUrl}${path}`;
  return (
    <Helmet htmlAttributes={{ lang }} title={title} titleTemplate={`%s | ${siteMetadata.title}`}>
      <link rel="canonical" href={url} />
      <meta name="description" content={metaDescription} />
      <meta name="image" content={imageUrl} />
      <meta name="keywords" content={keywords.join(', ')} />
      <meta property="og:description" content={metaDescription} />
      <meta property="og:image" content={imageUrl} />
      <meta property="og:title" content={fullTitle} />
      <meta property="og:type" content="website" />
      <meta property="og:url" content={url} />
      <meta name="twitter:card" content="summary_large_image" />
      <meta name="twitter:creator" content={siteMetadata.twitterUsername} />
      <meta name="twitter:description" content={metaDescription} />
      <meta name="twitter:image" content={imageUrl} />
      <meta name="twitter:title" content={fullTitle} />
    </Helmet>
  );
};

SEO.displayName = 'SEO';

export type SEOProps = {
  description?: string;
  keywords?: string[];
  lang?: string;
  path?: string;
  title: string;
};
export default SEO;
