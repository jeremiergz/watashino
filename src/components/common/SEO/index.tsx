import { graphql, useStaticQuery } from 'gatsby';
import React from 'react';
import { Helmet } from 'react-helmet-async';

const SEO = ({ description = '', keywords = [], lang = 'en', path = '', title }: SEO) => {
  const {
    banner,
    site: { siteMetadata },
  } = useStaticQuery(graphql`
    query SEOData {
      banner: file(relativePath: { eq: "cover.png" }) {
        childImageSharp {
          resize(width: 512) {
            src
          }
        }
      }
      site {
        siteMetadata {
          description
          siteUrl
          title
          twitterUsername
        }
      }
    }
  `);
  const metaDescription = description || siteMetadata.description;
  const siteUrl = siteMetadata.siteUrl;
  const imageUrl = `${siteUrl}${banner.childImageSharp.resize.src}`;
  const url = `${siteUrl}${path}`;
  return (
    <Helmet htmlAttributes={{ lang }} title={title} titleTemplate={`%s | ${siteMetadata.title}`}>
      <link rel="canonical" href={url} />
      <meta name="description" content={metaDescription} />
      <meta name="image" content={imageUrl} />
      <meta name="keywords" content={keywords.join(', ')} />
      <meta property="og:description" content={metaDescription} />
      <meta property="og:image" content={imageUrl} />
      <meta property="og:title" content={title} />
      <meta property="og:type" content="website" />
      <meta property="og:url" content={url} />
      <meta name="twitter:card" content="summary_large_image" />
      <meta name="twitter:creator" content={siteMetadata.twitterUsername} />
      <meta name="twitter:description" content={metaDescription} />
      <meta name="twitter:image" content={imageUrl} />
      <meta name="twitter:title" content={title} />
    </Helmet>
  );
};

SEO.displayName = 'SEO';

export default SEO;
