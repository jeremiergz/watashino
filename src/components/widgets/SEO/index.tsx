import { graphql, useStaticQuery } from 'gatsby';
import React from 'react';
import Helmet from 'react-helmet';

type Props = {
    description?: string;
    keywords?: string[];
    lang?: string;
    meta?: [];
    title: string;
};

function buildMeta({
    author,
    keywords,
    meta,
    metaDescription,
    title,
}: {
    author: string;
    keywords: string[];
    meta: [];
    metaDescription: string;
    title: string;
}) {
    return [
        { content: 'summary', name: 'twitter:card' },
        { content: 'website', property: 'og:type' },
        { content: author, name: 'twitter:creator' },
        { content: metaDescription, name: 'description' },
        { content: metaDescription, name: 'twitter:description' },
        { content: metaDescription, property: 'og:description' },
        { content: title, name: 'twitter:title' },
        { content: title, property: 'og:title' },
    ]
        .concat(keywords.length > 0 ? { content: keywords.join(`, `), name: `keywords` } : [])
        .concat(meta);
}

const SEOComponent = ({ description = '', keywords = [], lang = 'en', meta = [], title }: Props) => {
    const data = useStaticQuery<GraphQL.SeoQueryQuery>(graphql`
        query SEOQuery {
            site {
                siteMetadata {
                    title
                    description
                    author
                }
            }
        }
    `);
    const metaDescription = description || data.site.siteMetadata.description;
    const metadata = buildMeta({ author: data.site.siteMetadata.author, keywords, meta, metaDescription, title });
    return (
        <Helmet
            htmlAttributes={{ lang }}
            meta={metadata}
            title={title}
            titleTemplate={`%s | ${data.site.siteMetadata.title}`}
        />
    );
};

SEOComponent.displayName = 'SEO';

export { Props as SEOProps };
export default SEOComponent;
