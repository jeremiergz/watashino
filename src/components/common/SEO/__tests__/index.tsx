import * as Gatsby from 'gatsby';
import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import SEO from '../../SEO';

describe('components/widgets/SEO Test Suite', () => {
  beforeAll(() => {
    const useStaticQuery = jest.spyOn(Gatsby, 'useStaticQuery');
    useStaticQuery.mockImplementation(() => ({
      site: {
        siteMetadata: {
          title: 'SEO Test title',
          description: 'SEO Test description.',
          author: 'SEO Test author',
        },
      },
    }));
  });

  it('matches snapshot', () => {
    const jsx = <SEO description="SEO Test description" keywords={['test']} lang="en" title="SEO Test" />;
    expect(ReactTestRenderer.create(jsx).toJSON()).toMatchSnapshot();
  });

  it('matches snapshot without optional props', () => {
    const jsx = <SEO title="SEO Test" />;
    expect(ReactTestRenderer.create(jsx).toJSON()).toMatchSnapshot();
  });
});
