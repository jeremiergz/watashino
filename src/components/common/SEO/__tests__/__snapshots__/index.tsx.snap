// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/widgets/SEO Test Suite matches snapshot 1`] = `null`;

exports[`components/widgets/SEO Test Suite matches snapshot without optional props 1`] = `null`;
